baselines/flwr_baselines/publications/leaf/femnist/strategy.py

```diff
@@ -0,0 +1,88 @@
+"""FedAvg with the same clients used for both training and evaluation."""
+from typing import Callable, Dict, List, Optional, Tuple
+
+from flwr.common import (
+    EvaluateIns,
+    FitIns,
+    MetricsAggregationFn,
+    NDArrays,
+    Parameters,
+    Scalar,
+)
+from flwr.server import ClientManager
+from flwr.server.client_proxy import ClientProxy
+from flwr.server.strategy import FedAvg
+
+
+class FedAvgSameClients(FedAvg):
+    """FedAvg that samples clients for each round only once (the same clients
+    are used for training and testing round n)
+
+    It does not mean that the same client are used in each round. It used just the same clients
+    (with different parts of their data) in round i.
+
+    It assumes that there is no different function for evaluation - on_evaluate_config_fn
+    (it's ignored).
+    """
+
+    def __init__(
+        self,
+        *,
+        fraction_fit: float = 1.0,
+        fraction_evaluate: float = 1.0,
+        min_fit_clients: int = 2,
+        min_evaluate_clients: int = 2,
+        min_available_clients: int = 2,
+        evaluate_fn: Optional[
+            Callable[
+                [int, NDArrays, Dict[str, Scalar]],
+                Optional[Tuple[float, Dict[str, Scalar]]],
+            ]
+        ] = None,
+        on_fit_config_fn: Optional[Callable[[int], Dict[str, Scalar]]] = None,
+        on_evaluate_config_fn: Optional[Callable[[int], Dict[str, Scalar]]] = None,
+        accept_failures: bool = True,
+        initial_parameters: Optional[Parameters] = None,
+        fit_metrics_aggregation_fn: Optional[MetricsAggregationFn] = None,
+        evaluate_metrics_aggregation_fn: Optional[MetricsAggregationFn] = None,
+    ) -> None:
+        super().__init__(
+            fraction_fit=fraction_fit,
+            fraction_evaluate=fraction_evaluate,
+            min_fit_clients=min_fit_clients,
+            min_evaluate_clients=min_evaluate_clients,
+            min_available_clients=min_available_clients,
+            evaluate_fn=evaluate_fn,
+            on_fit_config_fn=on_fit_config_fn,
+            on_evaluate_config_fn=on_evaluate_config_fn,
+            accept_failures=accept_failures,
+            initial_parameters=initial_parameters,
+            fit_metrics_aggregation_fn=fit_metrics_aggregation_fn,
+            evaluate_metrics_aggregation_fn=evaluate_metrics_aggregation_fn,
+        )
```

Comment:
I just realised that this should work and be much more clean:

```suggestion
    def __init__(self, *args, **kwargs) -> None:
        super().__init__(*args, **kwargs)
```

Reply:
It's not a good coding practice. It takes less space but does not provide the necessary context for a potential user.

Reply:
I feel like in that case this is not really an issue, as when we usually write a custom strategy we don't even rewrite the __init__, which means that we don't consider it a problem to have those default parameters hidden.

Reply:
What I meant is that the arguments' names and their type won't be visible. Even though the code is shorter it does not make the code reusable more easily.

Reply:
In `flwr.server.strategy`, we repeat the arguments when deriving a new strategy from another existing strategy. In terms of consistency (or if we're considering to move this strategy into `flwr.server.strategy` at some point), it would make sense to repeat the args.